### DISCUSSION TITLE
How to pass a URL to firefox running within a private sandbox?

### DISCUSSION THREAD

**Discussion by oo-san** [netblue30/firejail#5082] [Q&A] [answered]
Hi all,
Related to #4670, FF94 has broken the workflow I used for opening links in different instances of Firefox running concurrently in different firejails with private home dirs. In short: when passing a URL to firefox for the initial launch, the URL is opened successfully in the respective sandbox. When firefox is already launched within multiple open sandbox, the most recently initialized sandbox captures the URL, (instead of the sandbox that I am trying to target).

I realize it's related to FF's recent change to fully using D-Bus for IPC, but I'm still trying to grok the details. If I set `dbus-user none` in firefox.local, then URLs get opened in the correct sandbox, but it kills FF's ability to open URLs in a currently-running instance of FF via the -new-tab argument. 

Here is a minimum reproducible example of what I'm trying to accomplish:

```
# create the directories for your private jails
mkdir -p testjails/jail1/.mozilla/firefox/main.default \
    testjails/jail2/.mozilla/firefox/main.default
# launch FF in jail 1
firejail --name=jail1 --profile=firefox --private=${HOME}/testjails/jail1 \
    firefox --profile ${HOME}/.mozilla/firefox/main.default -new-tab
# launch FF in jail 2 (new term)
firejail --name=jail2 --profile=firefox --private=${HOME}/testjails/jail2 \
    firefox --profile ${HOME}/.mozilla/firefox/main.default -new-tab
```
Expected behavior: Firefox opens in a new firejail (jail2).
Actual behavior: Firefox opens a new window in the first firejail (jail1)

Now, after adding `dbus-user none` to `${HOME}/.config/firejail/firefox.local` and re-run the same steps.
Expected behavior: Firefox opens in a new firejail (jail2).
Actual behavior: Firefox opens in a new firejail (jail2)!

Okay, all is good, now let's try passing a URL into firefox from jail1:
```
firejail --name=jail1 --profile=firefox --private=${HOME}/testjails/jail1 \                               
    firefox --profile ${HOME}/.mozilla/firefox/main.default -new-tab wikipedia.org

```
Expected behavior: Firefox opens wikipedia.org in jail1
Actual behavior: Firefox complains that it is already running but is not responding.

Okay, let's try joining the first sandbox instead:
```
firejail --join=jail1 \
    firefox --profile ${HOME}/.mozilla/firefox/main.default -new-tab wikipedia.org
```
Expected behavior: Firefox opens wikipedia.org in jail1
Actual behavior: Firefox complains that it is already running but is not responding.

It seems to me that setting `dbus-user` kills the ability for FF to know that it is already running and unsetting it allowes all running copies of FF to listen in on the same d-bus socket(?), with the most recently-opened copy capturing the new-tab request.

Is that all correct? Then how do I pass a URL into an already-open copy of FF running in a _specific_ sandbox?

Related issues/discussions:
#4670
#3919
#3471
#4756
#4532
https://github.com/glitsj16/firejail-handler-http (not relevant here I think, since I'm not trying to open a URL from _within_ a sandbox, I'm trying to pass a URL from outside the sandbox _into_ a _specific_ sandbox).

I've spent quite a long time trying to figure this out and it feels like I'm close.  Thanks for reading and helping me!

```
$ firefox --version
Mozilla Firefox 98.0.2
$ firejail --version
firejail version 0.9.68
```

**Comment by oo-san** [comment, marked as answer]
I think I found the issue.
Reading [this blog post](https://utcc.utoronto.ca/~cks/space/blog/unix/FirefoxDBusRemoteControl?showcomments) about FF's dbus implementation, it seems like the name of the bus is created deterministically from the profile name. [This still seems to be the case](https://searchfox.org/mozilla-central/rev/766773ca7e3495a3f1ba365f47dffefdea29e3ee/toolkit/components/remote/nsDBusRemoteClient.cpp#93).

The solution is simply to use a different profile name. Even though the profiles are in separate private homes mediated via the sandbox, they will still listen on the same bus if they are named the same. I was doing this to programmatically create disposable FF profiles to resist fingerprinting. I'm not sure if the profile name can be used as a fingerprinting vector, but at least opening external URLs works now.

Here is a working example:
```
# create the directories for your private jails
mkdir -p testjails/jail1/.mozilla/firefox/main1.default \
    testjails/jail2/.mozilla/firefox/main2.default
# launch FF in jail 1
firejail --name=jail1 --profile=firefox --private=${HOME}/testjails/jail1 \
    firefox --profile ${HOME}/.mozilla/firefox/main1.default -new-tab
# launch FF in jail 2 (new term)
firejail --name=jail2 --profile=firefox --private=${HOME}/testjails/jail2 \
    firefox --profile ${HOME}/.mozilla/firefox/main2.default -new-tab
```
Expected behavior: Firefox opens in a new firejail (jail2).
Actual behavior: Firefox opens in a new firejail (jail2)!

Passing URLs into the specific jailed FF also works now. I  suppose it should be possible to communicate directly with a sandboxed FF through dbus, as [findfox](https://github.com/ayosec/findfox) does, but I haven't tested this.

Still, it would be nice to know if I could re-name or re-direct a request sent to a dbus name either through or outside of firejail, to achieve the original intended results (that is: sending commands to FF instances in different jails, with different private homes, but the same profile name). If some script is able to peer at the profile directory and see the name, is that a fingerprinting vector?

**Reply by ghost** [reply]
That's a nice piece of research on this topic, congrats. After FF94 'broke' former routines several people got working on finding new ways to tackle the problem. I'll share a few notes I had lying around for future reference. Before doing so, one thing I notice is the somewhat unexpected use of the 'new tab' flag. According to my FF it's `--new-tab` and that option expects a URL. Might be a typo or GH eating stuff. Just wanted to point that out. IMO you don't need it in the initial setup of jails 1 & 2.

Alternatives:

(1) use xdg-desktop-portal to open links
- see [this](https://github.com/netblue30/firejail/issues/3881#issuecomment-805760626)
- xdg-open.c – A [xdg-open drop-in wrapper to make xdg-open work nicely with firejail](https://gist.github.com/rusty-snake/5104dc53ce3e52eef86cc34d359aa10e)

(2) use FF native remote debugging functionality
- [Mozilla info](https://developer.mozilla.org/en-US/docs/Tools/Remote_Debugging/Debugging_Firefox_Desktop)
- with tools like [chrome-remote-interface](https://github.com/cyrus-and/chrome-remote-interface) and [websocat](https://github.com/vi/websocat/) it's relatively easy to script together a custom way to deal with opening hyperlinks in FF - without relying on D-Bus:
```
enable remote debugging
    user_pref("devtools.chrome.enabled", true);
    user_pref("devtools.debugger.remote-enabled", true);

start FF with some extra flags for remote debugging
    --start-debugger-server [ws:][ <port> | <path>
        start the devtools server on a TCP port or Unix domain socket path (defaults to TCP port [6000])
        use WebSocket protocol if `ws:` prefix is specified
    --remote-debugging-port [<port>]
        start the FF remote agent, which is a low-level debugging interface based on the CDP protocol (defaults to listen on localhost:9222)

example
	$ firejail --name=jail1 --private=foo1 /usr/bin/firefox --remote-debugging-port 9222 --start-debugger-server ws:6000
	[...]
	DevTools listening on ws://localhost:9222/devtools/browser/405868c5-d1da-42a1-b5bb-822b6f0e3e13
	$ firejail --name=jail2 --private=foo2 /usr/bin/firefox --remote-debugging-port 9223 --start-debugger-server ws:6000
	[...]
	DevTools listening on ws://localhost:9223/devtools/browser/9d0b10d3-c517-4b77-95fb-1c9401255575

	Okay, all is good, now let's try passing a URL into firefox from jail1:

	$ chrome-remote-interface -n list | sort -r | grep -m 1 "9222/devtools/page/" | awk '{split($0,a,"\""); print a[4]}'
	ws://localhost:9222/devtools/page/2804824b-c47b-4c50-88c8-186d04424781
	$ echo "Page.navigate {\"url\":\"https://example.org\"}" | websocat --jsonrpc -q ws://localhost:9222/devtools/page/2804824b-c47b-4c50-88c8-186d04424781

	Okay, all is good, now let's try passing a URL into firefox from jail2:

	 $ chrome-remote-interface -p 9223 -n list | sort -r | grep -m 1 "9223/devtools/page/" | awk '{split($0,a,"\""); print a[4]}'
	ws://localhost:9223/devtools/page/b93adf3d-0195-45b5-a73c-1188a7d2d5b4
	$ echo "Page.navigate {\"url\":\"https://wikipedia.com\"}" | websocat --jsonrpc -q ws://localhost:9223/devtools/page/b93adf3d-0195-45b5-a73c-1188a7d2d5b4
```

> Still, it would be nice to know if I could re-name or re-direct a request sent to a dbus name either through or outside of firejail, to achieve the original intended results (that is: sending commands to FF instances in different jails, with different private homes, but the same profile name).

Maybe you can have a look at [dbus-action](https://github.com/bulletmark/dbus-action). Another nice tool for these kind of situations.

> If some script is able to peer at the profile directory and see the name, is that a fingerprinting vector?

With the profiles protected by the `private` home IMO it's quite tight. Rogue extensions or javascripts will always be adversaries in this cat & mouse game. Fingerprinting is a field of its own by now. Assume the worst and try to keep sensitive data away from web browsers as much as possible. But that's just my opinion, I have a pretty restrictive 'internet existence' :-P
